testing: BASH syntax - Add shellcheck to CI or as a bot

In #1168, a later comment came up that I was using camel case instead of snake case for a variable. As far as I can tell, we do not currently apply linting checks to shell scripts. Let's start that, either as part of our Kokoro pipeline or consider if we have an interest in using/enhancing a bot.

**Related issues:** https://github.com/GoogleCloudPlatform/nodejs-docs-samples/issues/1386

Shellcheck ([github](https://github.com/koalaman/shellcheck), [tryit](https://www.shellcheck.net/)) is the current front-runner in shell script linting. While it doesn't enforce variable name style it does enforce many other good practices.

### Usage Example

<details>
<summary>
Run with Docker:

```
docker run --rm -v "$PWD:/mnt" koalaman/shellcheck:stable \
    testing/kokoro/system_tests.sh
```

*Click to expand output*
</summary>

```
In system_tests.sh line 32:
for i in `find . -name go.mod`; do
         ^-------------------^ SC2044: For loops over find output are fragile. Use find -exec or a while read loop.
         ^-------------------^ SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean:
for i in $(find . -name go.mod); do


In system_tests.sh line 33:
  pushd `dirname $i` > /dev/null;
        ^----------^ SC2046: Quote this to prevent word splitting.
        ^----------^ SC2006: Use $(...) notation instead of legacy backticked `...`.
                 ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
  pushd $(dirname "$i") > /dev/null;


In system_tests.sh line 35:
    git diff go.mod | tee /dev/stderr | (! read)
                                           ^--^ SC2162: read without -r will mangle backslashes.


In system_tests.sh line 36:
    [ -f go.sum ] && git diff go.sum | tee /dev/stderr | (! read)
                                                            ^--^ SC2162: read without -r will mangle backslashes.


In system_tests.sh line 50:
export STORAGE_HMAC_ACCESS_KEY_ID="$($KOKORO_KEYSTORE_DIR/71386_golang-samples-kokoro-gcs-hmac-secret)"
       ^------------------------^ SC2155: Declare and assign separately to avoid masking return values.
                                     ^------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
export STORAGE_HMAC_ACCESS_KEY_ID="$("$KOKORO_KEYSTORE_DIR"/71386_golang-samples-kokoro-gcs-hmac-secret)"


In system_tests.sh line 51:
export STORAGE_HMAC_ACCESS_SECRET_KEY="$($KOKORO_KEYSTORE_DIR/71386_golang-samples-kokoro-gcs-hmac-id)"
       ^----------------------------^ SC2155: Declare and assign separately to avoid masking return values.
                                         ^------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
export STORAGE_HMAC_ACCESS_SECRET_KEY="$("$KOKORO_KEYSTORE_DIR"/71386_golang-samples-kokoro-gcs-hmac-id)"


In system_tests.sh line 67:
export GOLANG_SAMPLES_PROJECT_ID=$(gimmeproj -project golang-samples-tests lease $TIMEOUT);
       ^-----------------------^ SC2155: Declare and assign separately to avoid masking return values.


In system_tests.sh line 76:
trap "go clean -modcache; gimmeproj -project golang-samples-tests done $GOLANG_SAMPLES_PROJECT_ID" EXIT
                                                                       ^------------------------^ SC2064: Use single quotes, otherwise this expands now rather than when signalled.


In system_tests.sh line 102:
if [ -z ${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-} ]; then
        ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
if [ -z "${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-}" ]; then


In system_tests.sh line 108:
SIGNIFICANT_CHANGES=$(git --no-pager diff --name-only HEAD..master | egrep -v '(\.md$|^\.github)' || true)
                                                                     ^---^ SC2196: egrep is non-standard and deprecated. Use grep -E instead.


In system_tests.sh line 111:
CHANGED_DIRS=$(echo $SIGNIFICANT_CHANGES | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ')
                    ^------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
CHANGED_DIRS=$(echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ')


In system_tests.sh line 113:
if echo $SIGNIFICANT_CHANGES | tr ' ' '\n' | grep "^go.mod$" || [[ $CHANGED_DIRS =~ "testing" || $CHANGED_DIRS =~ "internal" ]]; then
        ^------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
if echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "^go.mod$" || [[ $CHANGED_DIRS =~ "testing" || $CHANGED_DIRS =~ "internal" ]]; then


In system_tests.sh line 132:
  TARGET=$(printf "./%s/... " $TARGET_DIRS)
                              ^----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
  TARGET=$(printf "./%s/... " "$TARGET_DIRS")


In system_tests.sh line 137:
if [ $GOLANG_SAMPLES_GO_VET ]; then
     ^--------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
if [ "$GOLANG_SAMPLES_GO_VET" ]; then


In system_tests.sh line 143:
  files=$(find $target_dir \( -exec [ -f {}/go.mod ] \; -prune \) -o -name "*.go" -print)
               ^---------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
  files=$(find "$target_dir" \( -exec [ -f {}/go.mod ] \; -prune \) -o -name "*.go" -print)


In system_tests.sh line 152:
    go vet $TARGET
           ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
    go vet "$TARGET"


In system_tests.sh line 157:
  for i in $(find $target_dir -name "go.mod")
           ^-- SC2044: For loops over find output are fragile. Use find -exec or a while read loop.
                  ^---------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
  for i in $(find "$target_dir" -name "go.mod")


In system_tests.sh line 159:
    mod="$(dirname $i)"
                   ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
    mod="$(dirname "$i")"


In system_tests.sh line 170:
2>&1 go test -timeout $TIMEOUT -v . $TARGET | tee $OUTFILE
                                    ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
2>&1 go test -timeout $TIMEOUT -v . "$TARGET" | tee $OUTFILE


In system_tests.sh line 174:
cat $OUTFILE | /go/bin/go-junit-report -set-exit-code > sponge_log.xml
    ^------^ SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.


In system_tests.sh line 181:
  gcloud auth activate-service-account --key-file $KOKORO_KEYSTORE_DIR/71386_kokoro-golang-samples-tests
                                                  ^------------------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
  gcloud auth activate-service-account --key-file "$KOKORO_KEYSTORE_DIR"/71386_kokoro-golang-samples-tests

For more information:
  https://www.shellcheck.net/wiki/SC2044 -- For loops over find output are fr...
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
  https://www.shellcheck.net/wiki/SC2064 -- Use single quotes, otherwise this...
```
</details>

## Considering a bot approach

If we think this should be approached as a bot, I've had https://github.com/googleapis/repo-automation-bots recommended, though there's also DPEBot.

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,5 +37,6 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
-      - run: ls
-      - run: shellcheck testing/kokoro/system_test.sh
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - run: find . -name "*.sh" -exec shellcheck {} \;

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -33,10 +33,5 @@ jobs:
       # based on command results. So, create a file if a build fails and check
       # if the file exists to set the right exit code.
       run: find . -name go.mod -execdir sh -c 'go build ./... || touch build_failed.txt' \; ; test ! -f build_failed.txt
-  shellcheck:
-    name: shellcheck
-    runs-on: ubuntu-latest
-    steps:
-      - name: Checkout code
-        uses: actions/checkout@v2
-      - run: find . -name "*.sh" -exec shellcheck {} \;
+    - name: shellcheck
+      run: find . -name "*.sh" -exec shellcheck {} \;

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,4 +37,4 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
-      - run: find . -name "*.sh" | xargs shellcheck
+      - run: find . -name '*.sh' | xargs shellcheck

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -33,3 +33,8 @@ jobs:
       # based on command results. So, create a file if a build fails and check
       # if the file exists to set the right exit code.
       run: find . -name go.mod -execdir sh -c 'go build ./... || touch build_failed.txt' \; ; test ! -f build_failed.txt
+  shellcheck:
+    name: shellcheck
+    runs-on: ubuntu-latest
+    steps:
+      - run: shellcheck **/*.sh

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,4 +37,5 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
+      - run: ls
       - run: shellcheck testing/kokoro/system_test.sh

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,5 +37,4 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
-      - run: pwd
-      - run: find . -name '*.sh' | xargs shellcheck
+      - run: shopt -s globstar && shellcheck **/*.sh

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,4 +37,4 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
-      - run: shellcheck **/*.sh
+      - run: find . -name "*.sh" | xargs shellcheck

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,4 +37,4 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
-      - run: shopt -s globstar && shellcheck **/*.sh
+      - run: shellcheck testing/kokoro/system_test.sh

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -37,4 +37,5 @@ jobs:
     name: shellcheck
     runs-on: ubuntu-latest
     steps:
+      - run: pwd
       - run: find . -name '*.sh' | xargs shellcheck

--- a/getting-started/devflowapp/deployment/sed_wrapper.sh
+++ b/getting-started/devflowapp/deployment/sed_wrapper.sh
@@ -3,6 +3,6 @@
 # Substitue values for environment variables in file $1
 
 echo "Substituting values for password and project ($PROJECT)"
-sed -e "s/_DB_PASSWORD/$_DB_PASSWORD/g" -i $1
-sed -e "s/PROJECT_ID/$PROJECT/g" -i $1
+sed -e "s/_DB_PASSWORD/$_DB_PASSWORD/g" -i "$1"
+sed -e "s/PROJECT_ID/$PROJECT/g" -i "$1"
 

--- a/getting-started/gce/startup-script.sh
+++ b/getting-started/gce/startup-script.sh
@@ -22,7 +22,7 @@ curl -s "https://storage.googleapis.com/signals-agents/logging/google-fluentd-in
 service google-fluentd restart &
 
 APP_LOCATION=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/app-location" -H "Metadata-Flavor: Google")
-gsutil cp $APP_LOCATION app.tar.gz
+gsutil cp "$APP_LOCATION" app.tar.gz
 tar -xzf app.tar.gz
 
 # Start the service included in app.tar.gz.

--- a/getting-started/gopher-run/cmd/training.sh
+++ b/getting-started/gopher-run/cmd/training.sh
@@ -1,3 +1,4 @@
+#! /bin/bash
 # Copyright 2019 Google LLC
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
@@ -35,19 +36,19 @@ while true; do
   DATE="$(date '+%Y%m%d_%H%M%S')"
   # Include date so JOB_ID is unique
   JOB_ID="${MODEL_NAME}_${DATE}"
-  gcloud beta ai-platform jobs submit training $JOB_ID \
+  gcloud beta ai-platform jobs submit training "$JOB_ID" \
   --master-image-uri=$IMAGE_URI --scale-tier=BASIC --job-dir=$JOB_DIR \
   -- \
   --preprocess --training_data_path=$TRAINING_DATA_PATH --objective=multi:softmax --num_class=4
 
   VERSION_NAME="V_${DATE}"
-  gcloud ai-platform versions create $VERSION_NAME \
+  gcloud ai-platform versions create "$VERSION_NAME" \
   --model ${MODEL_NAME} \
   --origin "${JOB_DIR}model/" \
   --runtime-version=1.11 \
   --framework ${FRAMEWORK} \
   --python-version=2.7
-  gcloud ai-platform versions set-default ${VERSION_NAME} --model=${MODEL_NAME}
+  gcloud ai-platform versions set-default "${VERSION_NAME}" --model="${MODEL_NAME}"
 
   # Repeate every 30 minutes
   sleep 1800

--- a/memorystore/redis/gce_deployment/deploy.sh
+++ b/memorystore/redis/gce_deployment/deploy.sh
@@ -18,7 +18,7 @@ fi
 
 if [ -z "$ZONE" ]; then
   ZONE=$(gcloud config get-value compute/zone -q)
-  echo $ZONE
+  echo "$ZONE"
 fi
 
 
@@ -36,8 +36,8 @@ gcloud compute instances create my-instance \
     --machine-type=g1-small \
     --scopes cloud-platform \
     --metadata-from-file startup-script=startup-script.sh \
-    --metadata gcs-bucket=$GCS_BUCKET_NAME,redis-host=$REDISHOST,redis-port=$REDISPORT \
-    --zone $ZONE \
+    --metadata gcs-bucket="$GCS_BUCKET_NAME",redis-host="$REDISHOST",redis-port="$REDISPORT" \
+    --zone "$ZONE" \
     --tags http-server
 
 gcloud compute firewall-rules create allow-http-server-8080 \

--- a/memorystore/redis/gce_deployment/startup-script.sh
+++ b/memorystore/redis/gce_deployment/startup-script.sh
@@ -4,6 +4,7 @@ set -ex
 
 # Talk to the metadata server to get the project id and location of application binary.
 PROJECTID=$(curl -s "http://metadata.google.internal/computeMetadata/v1/project/project-id" -H "Metadata-Flavor: Google")
+export PROJECTID
 GCS_BUCKET_NAME=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/gcs-bucket" -H "Metadata-Flavor: Google")
 REDISHOST=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/redis-host" -H "Metadata-Flavor: Google")
 REDISPORT=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/redis-port" -H "Metadata-Flavor: Google")

--- a/testing/docker/bootstrap.sh
+++ b/testing/docker/bootstrap.sh
@@ -24,6 +24,7 @@ GO111MODULE=off go get github.com/GoogleCloudPlatform/golang-samples/testing/gim
     github.com/jstemmer/go-junit-report \
     golang.org/x/tools/cmd/goimports
 
+(
 # Get the SDK tar and untar it.
 cd /tmp
 
@@ -40,5 +41,4 @@ rm $TARFILE
 
 ./google-cloud-sdk/bin/gcloud -q components update
 ./google-cloud-sdk/bin/gcloud -q components install app-engine-go
-
-cd -
+)

--- a/testing/kokoro/system_tests.sh
+++ b/testing/kokoro/system_tests.sh
@@ -39,15 +39,16 @@ TIMEOUT=60m
 
 # Also see trampoline.sh - system_tests.sh is only run for PRs when there are
 # significant changes.
-SIGNIFICANT_CHANGES=$(git --no-pager diff --name-only master..HEAD | egrep -v '(\.md$|^\.github)' || true)
+SIGNIFICANT_CHANGES=$(git --no-pager diff --name-only master..HEAD | grep -Ev '(\.md$|^\.github)' || true)
 # CHANGED_DIRS is the list of significant top-level directories that changed,
 # but weren't deleted by the current PR.
 # CHANGED_DIRS will be empty when run on master.
-CHANGED_DIRS=$(echo $SIGNIFICANT_CHANGES | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ' | xargs ls -d 2>/dev/null || true)
+CHANGED_DIRS=$(echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ' | xargs ls -d 2>/dev/null || true)
 
 # List all modules in changed directories.
 # If running on master will collect all modules in the repo, including the root module.
-GO_CHANGED_MODULES=$(find ${CHANGED_DIRS:-.} -name go.mod)
+# shellcheck disable=SC2086
+GO_CHANGED_MODULES="$(find ${CHANGED_DIRS:-.} -name go.mod)"
 # If we didn't find any modules, use the root module.
 GO_CHANGED_MODULES=${GO_CHANGED_MODULES:-./go.mod}
 # Exclude the root module, if present, from the list of sub-modules.
@@ -57,10 +58,10 @@ GO_CHANGED_SUBMODULES=${GO_CHANGED_MODULES#./go.mod}
 # Does not include static analysis checks.
 RUN_ALL_TESTS="0"
 # If this is a nightly test (not a PR), run all tests.
-if [ -z ${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-} ]; then
+if [ -z "${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-}" ]; then
   RUN_ALL_TESTS="1"
 # If the change touches a repo-spanning file or directory of significance, run all tests.
-elif echo $SIGNIFICANT_CHANGES | tr ' ' '\n' | grep "^go.mod$" || [[ $CHANGED_DIRS =~ "testing" || $CHANGED_DIRS =~ "internal" ]]; then
+elif echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "^go.mod$" || [[ $CHANGED_DIRS =~ "testing" || $CHANGED_DIRS =~ "internal" ]]; then
   RUN_ALL_TESTS="1"
 fi
 
@@ -68,20 +69,22 @@ fi
 # Do the easy stuff before running tests or reserving a project. Fail fast!
 set +x
 
-if [ $GOLANG_SAMPLES_GO_VET ]; then
+if [ "$GOLANG_SAMPLES_GO_VET" ]; then
   echo "Running 'goimports compliance check'"
   set -x
   diff -u <(echo -n) <(goimports -d .)
   set +x
   for i in $GO_CHANGED_MODULES; do
-    mod="$(dirname $i)"
-    pushd $mod > /dev/null;
+    mod=$(dirname "$i")
+    pushd "$mod" > /dev/null;
       # Fail if a dependency was added without the necessary go.mod/go.sum change
       # being part of the commit.
       echo "Running 'go.mod/go.sum sync check' in '$mod'..."
       set -x
       go mod tidy;
+      # shellcheck disable=SC2162
       git diff go.mod | tee /dev/stderr | (! read)
+      # shellcheck disable=SC2162
       [ -f go.sum ] && git diff go.sum | tee /dev/stderr | (! read)
       set +x
     popd > /dev/null;
@@ -98,8 +101,8 @@ if [ $GOLANG_SAMPLES_GO_VET ]; then
   # Recursive submodules are not supported.
   set +x
   for i in $GO_CHANGED_SUBMODULES; do
-    mod="$(dirname $i)"
-    pushd $mod > /dev/null;
+    mod=$(dirname "$i")
+    pushd "$mod" > /dev/null;
       echo "Running 'go vet' in '$mod'..."
       set -x
       go vet ./...
@@ -137,7 +140,8 @@ go install ./testing/sampletests
 # This is changed to a project-specific credential after a project is leased.
 export GOOGLE_APPLICATION_CREDENTIALS=$KOKORO_KEYSTORE_DIR/71386_kokoro-golang-samples-tests
 gimmeproj version;
-export GOLANG_SAMPLES_PROJECT_ID=$(gimmeproj -project golang-samples-tests lease $TIMEOUT);
+GOLANG_SAMPLES_PROJECT_ID=$(gimmeproj -project golang-samples-tests lease $TIMEOUT);
+export GOLANG_SAMPLES_PROJECT_ID
 if [ -z "$GOLANG_SAMPLES_PROJECT_ID" ]; then
   echo "Lease failed."
   exit 1
@@ -146,6 +150,7 @@ echo "Running tests in project $GOLANG_SAMPLES_PROJECT_ID";
 
 # Always return the project and clean the cache so Kokoro doesn't try to copy
 # it when exiting.
+# shellcheck disable=SC2064
 trap "go clean -modcache; gimmeproj -project golang-samples-tests done $GOLANG_SAMPLES_PROJECT_ID" EXIT
 
 set +x
@@ -183,19 +188,20 @@ runTests() {
   set +x
   echo "Running 'go test' in '$(pwd)'..."
   set -x
-  2>&1 go test -timeout $TIMEOUT -v ${1:-./...} | tee sponge_log.log
-  cat sponge_log.log | /go/bin/go-junit-report -set-exit-code > raw_log.xml
-  exit_code=$(($exit_code + $?))
+  2>&1 go test -timeout $TIMEOUT -v "${1:-./...}" | tee sponge_log.log
+  /go/bin/go-junit-report -set-exit-code < sponge_log.log > raw_log.xml
+  exit_code=$((exit_code + $?))
   # Add region tags tested to test case properties.
-  cat raw_log.xml | sampletests > sponge_log.xml
+  sampletests < raw_log.xml > sponge_log.xml
   rm raw_log.xml # No need to keep this around.
   set +x
 }
 
 if [[ $RUN_ALL_TESTS = "1" ]]; then
   echo "Running all tests"
+  # shellcheck disable=SC2044
   for i in $(find . -name go.mod); do
-    pushd "$(dirname $i)" > /dev/null;
+    pushd "$(dirname "$i")" > /dev/null;
       runTests
     popd > /dev/null;
   done
@@ -206,7 +212,7 @@ else
   runTests . # Always run root tests.
   echo "Running tests in modified directories: $CHANGED_DIRS"
   for d in $CHANGED_DIRS; do
-    mods="$(find $d -name go.mod)"
+    mods=$(find "$d" -name go.mod)
     # If there are no modules, just run the tests directly.
     if [[ -z "$mods" ]]; then
       pushd "$d" > /dev/null;
@@ -215,7 +221,7 @@ else
     # Otherwise, run the tests in all Go directories. This way, we don't have to
     # check to see if there are tests that aren't in a sub-module.
     else
-      goDirectories="$(find $d -name "*.go" -printf "%h\n" | sort -u)"
+      goDirectories="$(find "$d" -name "*.go" -printf "%h\n" | sort -u)"
       if [[ -n "$goDirectories" ]]; then
         for gd in $goDirectories; do
           pushd "$gd" > /dev/null;
@@ -230,8 +236,8 @@ fi
 # If we're running system tests, send the test log to the Build Cop Bot.
 # See https://github.com/googleapis/repo-automation-bots/tree/master/packages/buildcop.
 if [[ $KOKORO_BUILD_ARTIFACTS_SUBDIR = *"system-tests"* ]]; then
-  chmod +x $KOKORO_GFILE_DIR/linux_amd64/buildcop
-  $KOKORO_GFILE_DIR/linux_amd64/buildcop
+  chmod +x "$KOKORO_GFILE_DIR"/linux_amd64/buildcop
+  "$KOKORO_GFILE_DIR"/linux_amd64/buildcop
 fi
 
 exit $exit_code

--- a/testing/kokoro/trampoline.sh
+++ b/testing/kokoro/trampoline.sh
@@ -18,9 +18,9 @@ set -x
 
 date
 
-cd github/golang-samples
+cd github/golang-samples || exit 1
 
-SIGNIFICANT_CHANGES="$(git --no-pager diff --name-only master..HEAD | egrep -v '(\.md$|^\.github)' || true)"
+SIGNIFICANT_CHANGES="$(git --no-pager diff --name-only master..HEAD | grep -Ev '(\.md$|^\.github)' || true)"
 
 # If this is a PR with only insignificant changes, don't run any tests.
 if [[ -n ${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-} ]] && [[ -z "$SIGNIFICANT_CHANGES" ]]; then
@@ -28,11 +28,11 @@ if [[ -n ${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-} ]] && [[ -z "$SIGNIFICANT_CHANGE
   exit 0
 fi
 
-cd -
+cd - || exit 1
 
 function cleanup() {
-    chmod +x ${KOKORO_GFILE_DIR}/trampoline_cleanup.sh
-    ${KOKORO_GFILE_DIR}/trampoline_cleanup.sh
+    chmod +x "${KOKORO_GFILE_DIR}"/trampoline_cleanup.sh
+    "${KOKORO_GFILE_DIR}"/trampoline_cleanup.sh
     echo "cleanup";
 }
 trap cleanup EXIT

